sc-tools/sc-builder/src/gwf_parser.cpp

```diff
@@ -0,0 +1,171 @@
+#include "gwf_parser.hpp"
+
+std::string GwfParser::XmlCharToString(std::unique_ptr<xmlChar, XmlCharDeleter> const & ptr)
+{
+  return ptr ? std::string(reinterpret_cast<char const *>(ptr.get())) : "";
+}
+
+std::unique_ptr<xmlChar, XmlCharDeleter> GwfParser::GetXmlProp(xmlNodePtr node, char const * propName)
+{
+  return std::unique_ptr<xmlChar, XmlCharDeleter>(xmlGetProp(node, reinterpret_cast<xmlChar const *>(propName)));
+}
+
+void GwfParser::ProcessTag(xmlNodePtr child, std::unordered_map<std::string, std::string> & attributes)
+{
+  std::string const & tag = attributes["tag"];
+  if (tag == "node")
+  {
+    ProcessNode(child, attributes);
+  }
+  else if (tag == "pair" || tag == "arc")
+  {
+    ProcessPair(child, attributes);
+  }
+  else if (tag == "bus")
+  {
+    ProcessBus(child, attributes);
+  }
+  else if (tag == "contour")
+  {
+  }
+  else
+  {
+    throw std::runtime_error("Tag is not supported: " + tag);
+  }
+}
+
+void GwfParser::ProcessNode(xmlNodePtr node, std::unordered_map<std::string, std::string> & attributes)
+{
+  for (xmlNodePtr contentChild = node->children; contentChild != nullptr; contentChild = contentChild->next)
+  {
+    if (contentChild->type == XML_ELEMENT_NODE
+        && xmlStrEqual(contentChild->name, reinterpret_cast<xmlChar const *>("content")))
+    {
+      auto const contentTypeAttr = GetXmlProp(contentChild, "type");
+      auto const mimeTypeAttr = GetXmlProp(contentChild, "mime_type");
+      auto const fileNameAttr = GetXmlProp(contentChild, "file_name");
+      auto const contentData = std::unique_ptr<xmlChar, XmlCharDeleter>(xmlNodeGetContent(contentChild));
+
+      attributes["content_type"] = XmlCharToString(contentTypeAttr);
```

Comment:
Let implement hierarchy of sc.g-elements and incapsulate this fields in correspondent classes. You had better to do this, it can help to support this code in the future

Reply:
I propose this:

```
SCgElement
-> SCgNode
     -> SCgLink
     -> SCgContour
     -> SCgBus
-> SCgConnector
     -> SCgEdge
     -> SCgArc
```